src/collect/perf/mod.rs

```diff
@@ -0,0 +1,330 @@
+//! The Linux Perf trace collector.
+
+use super::PerfCollectorConfig;
+use crate::{
+    c_errors::PerfPTCError,
+    collect::{ThreadTraceCollector, TraceCollector},
+    errors::HWTracerError,
+    Trace,
+};
+use libc::{c_void, free, geteuid, malloc, size_t};
+use std::{convert::TryFrom, fs::File, io::Read, ptr, slice};
+
+extern "C" {
+    // collect.c
+    fn hwt_perf_init_collector(
+        conf: *const PerfCollectorConfig,
+        err: *mut PerfPTCError,
+    ) -> *mut c_void;
+    fn hwt_perf_start_collector(
+        tr_ctx: *mut c_void,
+        trace: *mut PerfTrace,
+        err: *mut PerfPTCError,
+    ) -> bool;
+    fn hwt_perf_stop_collector(tr_ctx: *mut c_void, err: *mut PerfPTCError) -> bool;
+    fn hwt_perf_free_collector(tr_ctx: *mut c_void, err: *mut PerfPTCError) -> bool;
+}
+
+const PERF_PERMS_PATH: &str = "/proc/sys/kernel/perf_event_paranoid";
+
+#[derive(Debug)]
+pub(crate) struct PerfTraceCollector {
+    config: PerfCollectorConfig,
+}
+
+impl PerfTraceCollector {
+    pub(super) fn new(config: PerfCollectorConfig) -> Result<Self, HWTracerError>
+    where
+        Self: Sized,
+    {
+        // Check for inavlid configuration.
+        fn power_of_2(v: size_t) -> bool {
+            (v & (v - 1)) == 0
+        }
+        if !power_of_2(config.data_bufsize) {
+            return Err(HWTracerError::BadConfig(String::from(
+                "data_bufsize must be a positive power of 2",
+            )));
+        }
+        if !power_of_2(config.aux_bufsize) {
+            return Err(HWTracerError::BadConfig(String::from(
+                "aux_bufsize must be a positive power of 2",
+            )));
+        }
+
+        Self::check_perf_perms()?;
+        Ok(Self { config })
+    }
+
+    fn check_perf_perms() -> Result<(), HWTracerError> {
```

Comment:
This could do with a doc string and, probably, a different name because it feels odd that `check` returns `Err`.

But, looking at it, I wonder why we don't just bundle all of this code into `new`?

Reply:
I will forget. So we can file an issue or do it as a commit in this PR. I think a commit in this PR is easiest, but it's up to you.

Reply:
Good point actually. OK let's nip these in the bud now.

Reply:
Works for me!

Reply:
Do you mind leaving this in a function? In the future we only need this check if we are going to collect a PT trace (and not e.g. some other kind of hardware trace). wrapping it around `if (PT) {` will make it a bit too-far indented for my taste.

Reply:
It's worse as it is because now *consumers* of this struct have to know if they're using PT or not! So I really think this should be merged.

Reply:
See #100 -- I plan to fix this soon.

Reply:
That feels like a separate issue to me. If there's a substantive argument against merging I'm all for it but "too much indentation" isn't very persuasive ;)

Reply:
Fixed in 19872e6